File: src/main/java/no/regnskap/mapper/RegnskapMapper.java
package no.regnskap.mapper;

import no.regnskap.generated.model.*;
import no.regnskap.model.persistance.RegnskapDB;
import no.regnskap.model.RegnskapXml;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;

public class RegnskapMapper {
    private static final String XML_TRUE_STRING = "J";

    public static List<RegnskapDB> mapFromXmlForPersistance(List<RegnskapXml> regnskapXml) {
        Map<String, RegnskapDB> toPersist = new HashMap<>();
        for (RegnskapXml xml : regnskapXml) {
            String key = xml.getHead().getOrgnr() + xml.getHead().getRegnaar();
            RegnskapDB mapped = toPersist.getOrDefault(key, new RegnskapDB());

            mapped.setOrgnr(xml.getHead().getOrgnr());
            mapped.setAarsregnskapstype(xml.getHead().getAarsregnskapstype());
            mapped.setJournalnr(xml.getHead().getJournalnr());
            mapped.setMottakstype(xml.getHead().getMottakstype());
            mapped.setOppstillingsplanVersjonsnr(xml.getHead().getOppstillingsplanVersjonsnr());
            mapped.setOrgform(xml.getHead().getOrgform());
            mapped.setRegnaar(xml.getHead().getRegnaar());
            mapped.setRegnskapstype(xml.getHead().getRegnskapstype());
            mapped.setValutakode(xml.getHead().getValutakode());

            mapped.setAvviklingsregnskap(booleanFromXmlData(xml.getHead().getAvviklingsregnskap()));
            mapped.setBistandRegnskapsforer(booleanFromXmlData(xml.getHead().getBistandRegnskapsforer()));
            mapped.setFeilvaloer(booleanFromXmlData(xml.getHead().getFeilvaloer()));
            mapped.setFleksiblePoster(booleanFromXmlData(xml.getHead().getFleksiblePoster()));
            mapped.setFravalgRevisjon(booleanFromXmlData(xml.getHead().getFravalgRevisjon()));
            mapped.setLandForLand(booleanFromXmlData(xml.getHead().getLandForLand()));
            mapped.setMorselskap(booleanFromXmlData(xml.getHead().getMorselskap()));
            mapped.setReglerSmaa(booleanFromXmlData(xml.getHead().getReglerSmaa()));
            mapped.setUtarbeidetRegnskapsforer(booleanFromXmlData(xml.getHead().getUtarbeidetRegnskapsforer()));
            mapped.setRevisorberetningIkkeLevert(booleanFromXmlData(xml.getHead().getRevisorberetningIkkeLevert()));

            mapped.setAvslutningsdato(localDateFromXmlDateString(xml.getHead().getAvslutningsdato()));
            mapped.setMottattDato(localDateFromXmlDateString(xml.getHead().getMottattDato()));
            mapped.setStartdato(localDateFromXmlDateString(xml.getHead().getStartdato()));

            mapped.setFields(RegnskapFieldsMapper.mapFieldsFromXmlData(mapped.getFields(), xml.getPosts()));

            toPersist.put(key, mapped);
        }

        return new ArrayList<>(toPersist.values());
    }

    private static boolean booleanFromXmlData(String trueOrFalse) {
        return XML_TRUE_STRING.equals(trueOrFalse);
    }

    public static Regnskap persistanceToGenerated(RegnskapDB persistanceDTO) {
        return new Regnskap()
            .id(persistanceDTO.getId())
            .avviklingsregnskap(persistanceDTO.isAvviklingsregnskap())
            .valuta(persistanceDTO.getValutakode())
            .oppstillingsplan(Regnskap.OppstillingsplanEnum.fromValue(persistanceDTO.getAarsregnskapstype().toLowerCase()))
            .revisjon(
                new Revisjon()
                    .ikkeRevidertAarsregnskap(persistanceDTO.isRevisorberetningIkkeLevert()))
            .regnskapsperiode(
                new Tidsperiode()
                .fraDato(persistanceDTO.getStartdato())
                .tilDato(persistanceDTO.getAvslutningsdato()))
            .regnkapsprinsipper(
                new Regnskapsprinsipper()
                    .smaaForetak(persistanceDTO.isReglerSmaa())
                    .regskapsregler(null)) //TODO Change to correct value
            .virksomhet(
                new Virksomhet()
                    .organisasjonsnummer(persistanceDTO.getOrgnr())
                    .organisasjonsform(persistanceDTO.getOrgform())
                    .morselskap(persistanceDTO.isMorselskap())
                    .levertAarsregnskap(true) // TODO Change to correct value
                    .navn(null)) // TODO Change to correct value
            .egenkapitalGjeld(persistanceDTO.getFields().getEgenkapitalGjeld())
            .eiendeler(persistanceDTO.getFields().getEiendeler())
            .resultatregnskapResultat(persistanceDTO.getFields().getResultatregnskapResultat());
    }

    private static LocalDate localDateFromXmlDateString(String dateString) {
        DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd");
        return LocalDate.parse(dateString, dateTimeFormatter);
    }

}

File: src/main/kotlin/no/regnskap/mapper/RegnskapMapper.kt
package no.regnskap.mapper

import no.regnskap.generated.model.Regnskap
import no.regnskap.generated.model.Regnskapsprinsipper
import no.regnskap.generated.model.Revisjon
import no.regnskap.generated.model.Tidsperiode
import no.regnskap.generated.model.Virksomhet
import no.regnskap.model.RegnskapDB
import no.regnskap.model.RegnskapFieldsDB
import no.regnskap.model.RegnskapXmlHead
import no.regnskap.model.RegnskapXmlWrap

import java.time.LocalDate
import java.time.format.DateTimeFormatter

private const val XML_TRUE_STRING = "J"
private val dateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd")

fun RegnskapXmlWrap.mapXmlListForPersistence(): List<RegnskapDB> {
    val toPersist: MutableMap<String, RegnskapDB> = HashMap()

    list.forEach {
        if(it.head != null) {
            val key = it.head.orgnr + it.head.regnaar
            val mapped = toPersist.getOrDefault(key, it.head.createRegnskapDB())

            mapped.fields = mapFieldsFromXmlData(mapped.fields, it.posts)
            toPersist[key] = mapped
        }
    }

    return ArrayList(toPersist.values)
}

private fun RegnskapXmlHead.createRegnskapDB(): RegnskapDB {
    val regnskapDB = RegnskapDB()

    regnskapDB.orgnr = orgnr
    regnskapDB.aarsregnskapstype = aarsregnskapstype
    regnskapDB.journalnr = journalnr
    regnskapDB.mottakstype = mottakstype
    regnskapDB.oppstillingsplanVersjonsnr = oppstillingsplanVersjonsnr
    regnskapDB.orgform = orgform
    regnskapDB.regnaar = regnaar
    regnskapDB.regnskapstype = regnskapstype
    regnskapDB.valutakode = valutakode

    regnskapDB.avviklingsregnskap = booleanFromXmlData(avviklingsregnskap)
    regnskapDB.bistandRegnskapsforer = booleanFromXmlData(bistandRegnskapsforer)
    regnskapDB.feilvaloer = booleanFromXmlData(feilvaloer)
    regnskapDB.fleksiblePoster = booleanFromXmlData(fleksiblePoster)
    regnskapDB.fravalgRevisjon = booleanFromXmlData(fravalgRevisjon)
    regnskapDB.landForLand = booleanFromXmlData(landForLand)
    regnskapDB.morselskap = booleanFromXmlData(morselskap)
    regnskapDB.reglerSmaa = booleanFromXmlData(reglerSmaa)
    regnskapDB.utarbeidetRegnskapsforer = booleanFromXmlData(utarbeidetRegnskapsforer)
    regnskapDB.revisorberetningIkkeLevert = booleanFromXmlData(revisorberetningIkkeLevert)

    regnskapDB.avslutningsdato = avslutningsdato.localDateFromXmlDateString()
    regnskapDB.mottattDato = mottattDato.localDateFromXmlDateString()
    regnskapDB.startdato = startdato.localDateFromXmlDateString()

    regnskapDB.fields = RegnskapFieldsDB()

    return regnskapDB
}


private fun booleanFromXmlData(trueOrFalse: String?): Boolean =
    XML_TRUE_STRING == trueOrFalse

fun RegnskapDB.mapPersistenceToGenerated(): Regnskap =
    Regnskap()
        .id(id?.toHexString())
        .avviklingsregnskap(avviklingsregnskap)
        .valuta(valutakode)
        .oppstillingsplan(Regnskap.OppstillingsplanEnum.fromValue(aarsregnskapstype.toLowerCase()))
        .revisjon(
            Revisjon()
                .ikkeRevidertAarsregnskap(revisorberetningIkkeLevert))
        .regnskapsperiode(
            Tidsperiode()
                .fraDato(startdato)
                .tilDato(avslutningsdato))
        .regnkapsprinsipper(
            Regnskapsprinsipper()
                .smaaForetak(reglerSmaa)
                .regnskapsregler(null)) //TODO Change to correct value
        .virksomhet(
            Virksomhet()
                .organisasjonsnummer(orgnr)
                .organisasjonsform(orgform)
                .morselskap(morselskap))
        .egenkapitalGjeld(fields.egenkapitalGjeld)
        .eiendeler(fields.eiendeler)
        .resultatregnskapResultat(fields.resultatregnskapResultat)

private fun String.localDateFromXmlDateString(): LocalDate =
    LocalDate.parse(this, dateTimeFormatter)

